zerver/lib/user_topics.py

```diff
@@ -244,27 +244,78 @@ def exclude_topic_mutes(
         # by not considering topic mutes outside the stream.
         query = query.filter(stream_id=stream_id)
 
-    rows = query.values(
+    excluded_topic_rows = query.values(
         "recipient_id",
         "topic_name",
     )
 
-    if not rows:
-        return conditions
-
     class RecipientTopicDict(TypedDict):
         recipient_id: int
         topic_name: str
 
-    def mute_cond(row: RecipientTopicDict) -> ClauseElement:
+    def topic_cond(row: RecipientTopicDict) -> ClauseElement:
         recipient_id = row["recipient_id"]
         topic_name = row["topic_name"]
         stream_cond = column("recipient_id", Integer) == recipient_id
         topic_cond = topic_match_sa(topic_name)
         return and_(stream_cond, topic_cond)
 
-    condition = not_(or_(*map(mute_cond, rows)))
-    return [*conditions, condition]
+    # Add this query later to reduce the number of messages it has to run on.
+    if excluded_topic_rows:
+        exclude_muted_topics_condition = not_(or_(*map(topic_cond, excluded_topic_rows)))
+        conditions = [*conditions, exclude_muted_topics_condition]
+
+    # Channel-level muting only applies when looking at views that
+    # include multiple channels, since we do want users to be able to
+    # browser messages within a muted channel.
+    if stream_id is None:
+        rows = Subscription.objects.filter(
+            user_profile=user_profile,
+            active=True,
+            is_muted=True,
+            recipient__type=Recipient.STREAM,
+        ).values("recipient_id")
+        muted_recipient_ids = [row["recipient_id"] for row in rows]
```

Comment:
Do we need to complete this query? I'd have sorta thought we might be able to just have the `UserTopic.objects.filter` query below include a filter on the `Stream` properties, and potentially save a database round-trip.

Reply:
Seems like yes, given how we use `muted_recipient_ids` separately below.